### PR TITLE
Issue 3888 - Custom Ticket Comments (with images)

### LINKED ISSUE #3888: Custom tickets - Comments with  images module
### Description
We need to have images on comments in custom tickets to have a good experience as a user.

### Goals
- Allow for multiple (potentially unlimited) images to be included
- Images can be attached via the "Attach image" icon
- Images can be attached by pasting them into the text field
- Images should be taken in consideration when replying.
- Clicking an image expands the image into a modal (same functionality as clicking the thumbnail in a custom ticket)

### Tasks
- [ ] 

### Related Resources
[MockUp](https://www.figma.com/file/3FjNEOdQbcHwzquVZxUqMl/3D-Repo?node-id=2444%3A19430&t=VGnFnPrq435bAoSU-0)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #3888

#### Description
Comments now support images.

Note a couple of things:
- Images can only be uploaded using the drag n drop system or by using the file icon. Adding them like you would in Markdown (`image`) will render as plain text;
- Editing a comment, should allow to also edit the images; however, the design for this is not ready yet. As such, editing a comment will preserve the existing images.

#### Test cases
- Create screenshots;
- Upload 1 or more images from file system;
- Drag and Drop images (the box should appear as soon as the dragging cursor reaches the new comment area) from file system or from other comments;
- Test the different images combinations (1, 2-5, 6+);
- Reply to a comment with 1 image;
- Reply to a comment with 1+ images;
- Try to expand the images in all the different combinations (including the upload phase)

### CODE REVIEW

**Comment by Amantini1997** [comment]
The following should also be addressed:

After submitting a comment, the chat should anchor to the bottom so you can see the last sent message. Currently, it maintains the current position.

image


### Make sure comment box is focused after clicking reply
After clciking reply, I have to click inside the comment box to type my reply. 

image



### Send button not visible

When the comment box is sticky, the send button, character count and image icon should be too
image

**Comment by The-Daniel** [comment]
image
This isn't related to images, but should there be some sort of indication that something is a markdown link? The "hello" in this image is actually a link to train times

**Comment by The-Daniel** [comment]
https://user-images.githubusercontent.com/62481250/219349706-161cd3cd-3dad-445f-8894-7884cbf84a09.mp4
I think there should perhaps be a max-height because you can upload very vertical images which take up lots of space

**Comment by The-Daniel** [comment]
image
Should there perhaps be a message explaining that videos (and other files) aren't supported?

**Comment by Amantini1997** [comment]
@mosment3DRepo @george-hadfield What do you think about the concerns Daniel raised?

**Comment by sanmont3drepo** [comment]
If I enter this 

\`\`\`
\`\`\` aaa
\`\`\`


the outcome is 
image

**Comment by sanmont3drepo** [comment]
probably nice to have: I think that if I have a comment in edit mode and I create a new comment, that edit mode  should stop

File: frontend/src/v5/ui/controls/inputs/inputController.component.tsx [hunk 1]
```diff
@@ -66,7 +67,7 @@ export const InputController = <T,>({
 					field.onBlur();
 					onBlur?.();
 				}}
-				inputRef={ref}
+				inputRef={inputRef || ref}
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/controls/inputs/inputController.component.tsx:70]: What happened here?
I think ref  can only be passed using fowardRef (it doesnt show as a prop)

File: frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx [hunk 2]
```diff
@@ -44,17 +44,22 @@ const ThumbHorizontal = styled.div<{ variant: ScrollbarVariant }>`
 	border-radius: 3px;
 	height: 6px;
 	background-color: ${({ variant }) => getBackgroundColor(variant)};
+	overflow-x: hidden;
 `;
-
-export const ScrollbarWrapper = styled(Scrollbars).attrs(
-	({ variant, $hidehorizontal }: { variant: ScrollbarVariant, $hidehorizontal: boolean }) => ({
+export type ScrollAreaProps = ScrollbarProps & {
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx:49]: why the change of scrolllArea.component?

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/basicComment/basicComment.component.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+export type BasicCommentProps = {
+	images?: string[];
+	children?: any;
+	className?: string;
+	message?: string;
+	commentAge: string;
+	history?: TicketCommentHistoryBlock[];
+	metadata?: TicketCommentReplyMetadata;
+	isCurrentUserComment?: boolean;
+	author?: string;
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/basicComment/basicComment.component.tsx:36]: Should author be required?

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/createCommentBox/createCommentBox.component.tsx [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+						))}
+					</Images>
+				</ScrollArea>
+			)}
+			{erroredImages.length > 0 && erroredImages.map(({ name }) => (
+				<ErroredImageMessage>
+					<strong>{name} </strong>
+					<FormattedMessage
+						id="customTicket.comments.images.error"
+						defaultMessage="is too big. 1GB limit."
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/createCommentBox/createCommentBox.component.tsx:220]: Should the filesize here be taken from the config?

Comment (The-Daniel) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/createCommentBox/createCommentBox.component.tsx:220]: And is it not possible for other errors to occur?

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Comment by Amantini1997** [comment]
> If I enter this
> 
> ``` ``` aaa ```
> 
> the outcome is image
> 
> in edit mode: image

I will change the "sanitisation" process so that all the message will be fully processed by the markdown library. This will fix errors of this kind, but also get rid of multi-linebreaks not defined using `<br/>`

**Comment by Amantini1997** [comment]
> probably nice to have: I think that if I have a comment in edit mode and I create a new comment, that edit mode should stop

In Teams, when you are editing a comment and start typing in the input box at the bottom, nothing happens to the comment currently being edited. But @mosment3DRepo @sanmont3drepo let me know if you want that behaviour to change

File: frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	isFederation,
+	commentId,
+	comment,
+}: UpdateCommentAction) {
+	try {
+		const updateModelComment = isFederation
+			? API.TicketComments.updateFederationComment
+			: API.TicketComments.updateContainerComment;
+
+		yield updateModelComment(teamspace, projectId, modelId, ticketId, commentId, parseMessageAndImages(comment));
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts:127]: The comment object being sent to the backend, follows this structure
```ts
{
  images: string[],
  message: string,
  ...
}
```

When the message is empty, instead of sending `{ message: '', ... }`, the `message` property should be deleted
Something similar goes for the images, if there are no images, instead of sending `{ images: [], ... }`, the `images` property should be deleted.

So before sending the new (or the updated) comment to the backend, it shall be processed as described above.

One last operation is to strip the 64BasePrefix on the images (if present).

Regarding the snippet of code that this comment refers to, you can see why it needed by doing this:
- send a comment with a message and at least one image;
- update the message so that it is empty;
- save the comment;

This will succeed, the comment will be correctly updated. But if you were to remove the `parseMessageAndImages` (line 127), the following steps will result in an error because the message property is being sent to the backend as an empty string instead of being removed from the comment object

File: frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx (hunk 2 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx:49]: The scrollArea component was not doing anything other than the setting a defaultvalue for `hideHorizontal` and `variant`

Comment (The-Daniel) [reply frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx:49]: I'm a bit confused because it seems to work as it currently is, but shouldn't the ternary operator in renderThumbHorizontal be the other way around?

File: frontend/src/v5/ui/controls/inputs/inputController.component.tsx (hunk 1 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/controls/inputs/inputController.component.tsx:70]: https://mui.com/material-ui/api/input/

inputRef is a prop accepted by MUI input. It allows to pass a ref down to the actual input component without using `forwardRef` several times across the children before reaching the input.

Here you can see that the inputRef prop reaches `InputBase` and after that it is treated like an actual ref, so no longer visible in children as a prop
image

File: frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx (hunk 2 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/controls/scrollArea/scrollArea.styles.tsx:49]: When I noticed that, I got pretty confused myself. However, this PR was already big enough, so I didn't want to refactor it here also risking to break other components. Plus, the golden rule states "if it works, don't touch it" lol

File: frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+		onError(error);
+		yield put(DialogsActions.open('alert', {
+			currentActions: formatMessage(
+				{ id: 'comments.create.error', defaultMessage: 'trying to create the comment for {model} ticket' },
+				{ model: isFederation ? 'federation' : 'container' },
+			),
+			error,
+			details: formatMessage({
+				id: 'comments.create.error.details',
+				defaultMessage: 'If reloading the page doesn\'t work please contact support',
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts:107]: Good shout

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/basicComment/basicComment.component.tsx (hunk 3 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/basicComment/basicComment.component.tsx:36]: Nope, if you look at line 58, there is a conditional rendering, meaning the author may be empty.
This could happen if the user is from currentUser, if it deleted, or if it from otherUser, but it is not the first contiguous message from that author

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/createCommentBox/createCommentBox.component.tsx (hunk 4 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/createCommentBox/createCommentBox.component.tsx:220]: Error should be known in order to be displayed properly. The only 2 errors that I can think of are:
1. The file is too big
2. The file is corrupted: the extension asserts to be an allowed one, but the in the backend a check is performed and that image is actually not valid (e.g. webp)

The second error triggers a modal explaining the issue, but it not possible to detect what file is responsible for the error since the error response does not contain any information about it.

Regarding the first error, for `editProfileAvatar` (`src\v5\validation\userSchemes\validators.ts`) the size is hardcoded. There is an helper function that maps the size value to a string (`formatInfoUnit`), but the output is "104.9 MB". If @sanmont3drepo is ok with that, I can use that function instead

File: frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	try {
+		const fetchModelComments = isFederation
+			? API.TicketComments.fetchFederationComments
+			: API.TicketComments.fetchContainerComments;
+		const { comments } = yield fetchModelComments(teamspace, projectId, modelId, ticketId);
+		const richComments = sortBy(comments, 'createdAt').map(({ createdAt, updatedAt, ...comment }) => ({
+			...comment,
+			createdAt: new Date(createdAt),
+			updatedAt: new Date(updatedAt),
+		}));
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/tickets/comments/ticketComments.sagas.ts:49]: I tested it on a ticket with over 2k comments. The performance is pretty much the same. 

To test it, you can run this script which will give you an idea on how much it takes to perform a `new Date` operation (n = 2,000 -> 2ms)
```ts
const timeDate = (n) => {
    const title = `date ${n} times`
    console.time(time);
    Array(n).fill(0).forEach(() => new Date(1632821119000));
    console.timeEnd(time);
}
```

File: frontend/src/v5/ui/controls/inputs/inputController.component.tsx (hunk 1 above)
Comment (sanmont3drepo) [reply frontend/src/v5/ui/controls/inputs/inputController.component.tsx:70]: I think we should have one way of using it either ref or inputRef, but not both.

Comment (Amantini1997) [reply frontend/src/v5/ui/controls/inputs/inputController.component.tsx]: The ref is passed an actual ref now

**Review by sanmont3drepo — APPROVED** [review]

**Review by sanmont3drepo — DISMISSED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/commentMarkDown/commentMarkDown.tsx [hunk 8]
```diff
@@ -15,14 +15,17 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import { desanitiseMessage } from '@/v5/store/tickets/comments/ticketComments.helpers';
 import ReactMarkdown from 'react-markdown';
 import { CommentMessage } from '../basicComment/basicComment.styles';
 
-const SanitisedImage = ({ src }) => (<>![image]({src})</>);
+const ImageMarkdown = ({ src }) => (<>![image]({src})</>);
+const CodeMarkdown = ({ value }) => (<>{desanitiseMessage(value)}</>);
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/commentsPanel/comment/commentMarkDown/commentMarkDown.tsx]: There is a syntax in markdown that allows multiline code. The syntax is the following:
\`\`\``<code language, e.g. ts or scss>`
code to be highlight
\`\`\`
</code>


In your case, the `I am a secret code` is just the language that you are saying should be used to tokenise and lint the code inside the backticks. As such, it makes sense it does not get displayed.

When I am not sure what the output should be, I like to use this [website](https://markdownlivepreview.com/) as reference. It's an MD previewer
image

**Review by The-Daniel — DISMISSED** [review]

**Comment by george-hadfield** [comment]
### Circle background should be transparent for gallery icons
Currently has blue fill

image

**Comment by george-hadfield** [comment]
Should be able to use left and right keys to scroll through images in gallery
image

**Review by george-hadfield — CHANGES_REQUESTED** [review]
A few tweaks

**Comment by Amantini1997** [comment]
> ### Circle background should be transparent for gallery icons
> Currently has blue fill
> 
> image

I did it because when the device screen is not large enough and the arrows overlap the images, they are easier to see (especially on screenshots, as the viewer background colour is very light). Below is an example. Let me know if I should revert the fill to be transparent nonetheless

**with fill**
image

**without fill**
image

**Comment by Amantini1997** [comment]
> "Drag and Drop images (the box should appear as soon as the dragging cursor reaches the new comment area) from file system or from other comments;"
> 
> This isn't working for me. No image is stored after dropping on the drag and drop area

Do you drop it over the dashed area (i.e. the dashed border becomes a solid line and the background becomes teal)?

**Comment by george-hadfield** [comment]
> r the dashed area (i.e. the dashed

[screen-capture (6).webm](https://user-images.githubusercontent.com/78047794/222402014-699b661c-7e2a-4d81-9368-df17d4a26845.webm)

Using **Chrome Version 110.0.5481.177 (Official Build) (x86_64)**

**Comment by george-hadfield** [comment]
> > ### Circle background should be transparent for gallery icons
> > Currently has blue fill
> > image
> 
> I did it because when the device screen is not large enough and the arrows overlap the images, they are easier to see (especially on screenshots, as the viewer background colour is very light). Below is an example. Let me know if I should revert the fill to be transparent nonetheless
> 
> **with fill** image
> 
> **without fill** image

In this case, the image should scale down with the browser to leave enough space for the circles to be left and right of the image without overlapping. Is that ok?

**Review by george-hadfield — CHANGES_REQUESTED** [review]
Replied to your comments

**Comment by george-hadfield** [comment]
### Remove jump animation with drag and drop in comments

[screen-capture (7).webm](https://user-images.githubusercontent.com/78047794/222417791-26dfa9f2-a17f-472d-9420-fec12648ff96.webm)

**Review by george-hadfield — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by The-Daniel — APPROVED** [review]
